streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -16,128 +16,94 @@
+                                       final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients,
+                                       final Map<UUID, ClientState> clientStates,
+                                       final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                       final int maxWarmupReplicas) {
+        boolean warmupReplicasAssigned = false;
+
+        final ValidClientsByTaskLoadQueue clientsByTaskLoad =
+            new ValidClientsByTaskLoadQueue(
+                clientStates,
+                (client, task) -> taskIsCaughtUpOnClient(task, client, tasksToCaughtUpClients)
+            );
 
-        final Map<TaskId, UUID> taskToDestinationClient = new HashMap<>();
-        for (final Map.Entry<UUID, List<TaskId>> clientEntry : balancedStatefulActiveTaskAssignment.entrySet()) {
-            final UUID destination = clientEntry.getKey();
-            for (final TaskId task : clientEntry.getValue()) {
-                taskToDestinationClient.put(task, destination);
+        final SortedSet<TaskMovement> taskMovements = new TreeSet<>(
+            (movement, other) -> {
+                final int numCaughtUpClients = tasksToCaughtUpClients.get(movement.task).size();
+                final int otherNumCaughtUpClients = tasksToCaughtUpClients.get(other.task).size();
+                if (numCaughtUpClients != otherNumCaughtUpClients) {
+                    return numCaughtUpClients - otherNumCaughtUpClients;
+                } else {
+                    return movement.task.compareTo(other.task);
+                }
             }
+        );
+
+        for (final Map.Entry<UUID, List<TaskId>> assignmentEntry : statefulActiveTaskAssignment.entrySet()) {
+            final UUID client = assignmentEntry.getKey();
+            final ClientState state = clientStates.get(client);
+            for (final TaskId task : assignmentEntry.getValue()) {
+                if (taskIsCaughtUpOnClient(task, client, tasksToCaughtUpClients)) {
+                    state.assignActive(task);
+                } else {
+                    final TaskMovement taskMovement = new TaskMovement(task,  client);
+                    taskMovements.add(taskMovement);
+                }
+            }
+            clientsByTaskLoad.offer(client);
         }
 
-        int remainingAllowedWarmupReplicas = maxWarmupReplicas;
-        final List<TaskMovement> movements = new LinkedList<>();
-        for (final Map.Entry<UUID, List<TaskId>> sourceClientEntry : statefulActiveTaskAssignment.entrySet()) {
-            final UUID source = sourceClientEntry.getKey();
+        int remainingWarmupReplicas = maxWarmupReplicas;
+        for (final TaskMovement movement : taskMovements) {
+            final UUID leastLoadedClient = clientsByTaskLoad.poll(movement.task);
+            if (leastLoadedClient == null) {
+                throw new IllegalStateException("Tried to move task to caught-up client but none exist");
```

Comment:
Sure, that sounds even better. We just need to bear in mind the difference in audience. Exception messages need to be worded in a way that will make sense to users. Nothing wrong with that, though.

Reply:
I guess the way I thought about it was:
1. the `clientsByTaskLoad` queue returns the (next least loaded) caught-up client for a given task 
2. the `taskMovements` set only contains tasks with caught-up clients
3. therefore, every task in `taskMovements` has at least one client

If I'm following this discussion correctly, it seems like point 2. is the root cause of confusion. Is that a fair summary? I'll definitely rename `taskIsCaughtUpOnClient`, thanks for pointing that out. Would it help to leave a comment on the actual `TaskMovement` class clarifying that a task to be moved necessarily has at least one caught-up client (because otherwise why would we try to move it), and/or rename `taskMovements` (suggestions welcome) ?

Reply:
I'm having trouble thinking of a way to make the exception message more clear without just starting to name variable names and describe the literal code, which doesn't seem appropriate for an exception message. Does anyone have any suggestions? Maybe something like `scheduled a task to be moved and assigned to a caught-up client, but no caught up clients were found`? That just seems like a more verbose way to say the same thing 😕

Reply:
Actually, I think I was saying something different. There is not a guarantee that any client is caught up, which is why the name `taskIsCaughtUpOnClient` is misleading. If that name is accurate, then it looked like the IllegalStateException would actually get thrown all the time, whenever there is no caught-up client.
So the subtlety is that we _do_ get back a client when there is no caught-up client. In that case, the predicate falls back to giving us all the clients. Hence the name I recommended. I do think this is all fairly esoteric from the user's perspective, which may be why I thought first of leaving a comment.
I also can't think of what to say in the exception. What I said before references internal variables, which is not appropriate. But there also doesn't really seem to be a way to say "this is why this happened" because the whole point of ISE is that we have no idea why it would happen.
How about you just take as feedback that this is one of the subtlest things I've seen in a while, and make some attempt to make it less mind-blowing (maybe just renaming the method is good enough).

Worst-case scenario, if we come back in a year and again get confused, we'll have another chance to clarify the code. This feedback is just about trying to prevent that future, but we can only do our best.

Reply:
But we only put a task in the `taskMovements` set if we know it has a caught-up client (that isn't its current client). The inverse is also true; if a task has no caught-up clients, it won't be in `taskMovements`. So any task we get while looping through `taskMovements` necessarily has at least one caught-up client, right?
I agree here that `taskIsCaughtUpOnClient` was misleading, and have renamed it (just haven't pushed the latest changes yet).

Reply:
Let's say we rename it to `taskIsCaughtUpOnClientOrNoCaughtUpClientsExist` Would you still feel like this is a subtle issue worth leaving a comment for? Or was that the sole source of confusion?

Reply:
I think that would be good enough. Thanks!

Going back to your prior comment,

> But we only put a task in the taskMovements set if we know it has a caught-up client (that isn't its current client). The inverse is also true; if a task has no caught-up clients, it won't be in taskMovements. So any task we get while looping through taskMovements necessarily has at least one caught-up client, right?

I feel like I'm just missing something here. It looks to me like the only precondition for adding a task to `taskMovements` is that it is _not_ caught up on the destination. Why does that imply that we know it _is_ caught up on another?

Reply:
```if (taskIsCaughtUpOnClientOrNoCaughtUpClientsExist(task, client, tasksToCaughtUpClients)) {
                    state.assignActive(task);
                } else {
                    final TaskMovement taskMovement = new TaskMovement(task, client);
                    taskMovements.add(taskMovement);
                }
```
with the renaming, this block now looks like this. If no caught up clients exist, we just assign it to the ClientState right away; otherwise, if and only if `taskIsCaughtUpOnClientOrNoCaughtUpClientsExist` returns false, do we add to `taskMovements`.

So by (I think it's called deMorgan's Law?):
!(caughtUpOnClient | noCaughtUpClientsExist) 
== !caughtUpOnClient & !noCaughtUpClientsExist 
== notCaughtUpOnClient & caughtUpClientsExist

I also feel like I might be missing your point, but I hope we can get somewhere when the method is more appropriately named 🙂

Reply:
Haha, ok, now I get it. See? I knew renaming the method was a good idea. I think we can definitely chalk this one up as a win for me ;)

Nice work busting out the `∴` character.

Reply:
Yeah I felt pretty fancy with the `∴`. 

Hope that didn't come off as too condescending, it's just been a while since I got to practice my logical proofs and I couldn't resist 😄

Reply:
Hah! No, not offended at all. I think the old phrase is "Flawless Victory."